rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+2. Set headers to prevent caching entirely
+
+#### Changes to `vcl_deliver`
+
+When delivering a response to the user, set a new cookie with a new
+expiration time, and disable shared caches outside of Fastly (both
+Fastly and the user's browser can still cache the page) if the
+response depended on the session:
+
+```vcl
+if (req.http.GOVUK-End-Session == "true") {
+    set req.http.GOVUK-Session = "";
+} else if (req.http.GOVUK-Session-ID) {
+  set req.http.GOVUK-Session-Expires = strftime({"%s"}, time.add(now, 1800s));
+  set req.http.GOVUK-Session-Signature = digest.hmac_sha256(table.lookup(env, "session_key"), req.http.GOVUK-Session-ID + "." + req.http.GOVUK-Session-Expires));
+  set req.http.GOVUK-Session = digest.base64url_nopad(req.http.GOVUK-Session-ID) + "." + digest.base64url_nopad(req.http.GOVUK-Session-Expires) + "." + digest.base64url_nopad(req.http.GOVUK-Session-Signature);
+}
+
+if (req.http.GOVUK-Session) {
+  add resp.http.Set-Cookie = "govuk_session=" + req.http.GOVUK-Session + "; secure; httponly; samesite=strict; max-age=1800; path=/";
+}
+
+if (resp.http.Vary ~ "GOVUK-Session-ID") {
+  unset resp.http.Vary:GOVUK-Session-ID;
+  set resp.http.Cache-Control:private = "";
+}
+```
+
+The `1800` here (30 minutes) is just an example.  We will work out an
+appropriate session duration with product and security input.
+
+### Create a new app to manage the auth process
+
+It's weird to have the login and logout controllers for GOV.UK as a
+whole located under `/transition-check`.
+
+We will create a new app—called account-frontend, which will live on a
+new machine class called personalisation—to serve those controllers
+and to handle OAuth interactions with the GOV.UK Account
+system. Signing in will set the `GOVUK-Session-ID` response header,
+signing out will set the `GOVUK-End-Session` response header.
+
+This app will serve these public endpoints:
+
+- `GET /sign-in`: initiates the OAuth flow with the accounts system
+  and sends the user on a consent journey.  Accepts these parameters:
+
+    - `_ga`: cross-domain tracking parameter to pass to the accounts
+      domain (optional)
+    - `redirect_path`: path on GOV.UK to redirect back to after
+      authenticating (optional, default: `/`)
+    - `state`: see below (optional)
+
+- `GET /sign-in/callback`: where the accounts system sends the user
+  back to.  Sets the `GOVUK-Session-ID` response header and stores the
+  returned OAuth access and refresh tokens in the app's database, if
+  the user successfully signed in, and redirects the user back to the
+  redirect_path.
+
+- `GET /sign-out`: sets the `GOVUK-End-Session` response header,
+  deletes the OAuth tokens from the app's database, and redirects the
+  user to `/`.
+
+The app will also serve these internal endpoints:
+
+- `GET /api/attributes`: looks up and returns some attributes from the
+  user's account.  Accepts these parameters:
+
+    - `session_id`: the value of the `GOVUK-Session-ID` header
```

Comment:
I'm assuming a user is a record that references their GOV.UK account user id - so one user per account. And that you use the session-id to determine the account. Can a user have multiple valid session-ids at a time?

Reply:
If a user is logged in from multiple devices, then they'll have multiple sessions.  I'll add a note saying that the session ID is a random value generated a log in.